nixos/modules/services/backup/sanoid.nix

```diff
@@ -0,0 +1,163 @@
+{ config, stdenv, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.sanoid;
+
+  percentType = (types.ints.between 0 100) // {
+    name = "Percent";
+  };
+
+  commonOptions = {
+    hourly = mkOption {
+      description = "Number of hourly snapshots";
+      type = types.nullOr types.ints.unsigned;
+      default = null;
+    };
+
+    daily = mkOption {
+      description = "Number of daily snapshots";
+      type = types.nullOr types.ints.unsigned;
+      default = null;
+    };
+
+    monthly = mkOption {
+      description = "Number of monthly snapshots";
+      type = types.nullOr types.ints.unsigned;
+      default = null;
+    };
+
+    yearly = mkOption {
+      description = "Number of yearly snapshots";
+      type = types.nullOr types.ints.unsigned;
+      default = null;
+    };
+
+    autoprune = mkOption {
+      description = "Whether to automatically prune old snapshots.";
+      type = types.nullOr types.bool;
+      default = null;
+    };
+
+    autosnap = mkOption {
+      description = "Whether to automatically take snapshots.";
+      type = types.nullOr types.bool;
+      default = null;
+    };
+
+    extraConfig = mkOption {
+      description = "Extra configuration for this template/dataset";
```

Comment:
As I'm describing in NixOS/rfcs#42, introducing such `extraConfig` options has problems, so I'd rather have this PR use the approach described there. This means:
- Changing this to `settings = mkOption { type = types.attrsOf (types.nullOr (types.oneOf [ int bool str ])); ... }`
- Setting all other options in the `config` section of the (sub)module like `settings.hourly = mkDefault cfg.hourly;` instead
- Generating the config file *only* from the `cfg.settings` values. You might be able to use `lib.generators.toINI` for this

If you want to I could implement this for you

Reply:
I implemented this; let me know if it looks good to you.